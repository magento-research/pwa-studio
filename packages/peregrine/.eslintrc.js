const config = {
    parser: 'babel-eslint',
    extends: ['@magento'],
    rules: {
        "no-undef": "off"
    }
};

module.exports = config;
