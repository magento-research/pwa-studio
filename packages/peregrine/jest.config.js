module.exports = {
    displayName: 'Peregrine',
    setupFiles: ['<rootDir>/scripts/shim.js', '<rootDir>/scripts/fetch-mock.js']
};
