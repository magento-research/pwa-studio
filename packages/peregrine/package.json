{
  "name": "@magento/peregrine",
  "version": "3.0.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "build": " ",
    "build:ci": " ",
    "build:cjs": " ",
    "build:dev": " ",
    "clean": " ",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "storybook:build": "build-storybook -c .storybook -o storybook-dist",
    "watch": " "
  },
  "repository": "github:magento-research/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento-research/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine#readme",
  "devDependencies": {
    "@storybook/react": "~4.1.13",
    "intl": "~1.2.5",
    "react": "~16.8.3",
    "react-apollo": "~2.5.3",
    "react-dom": "~16.8.3",
    "react-router-dom": "~5.0.0",
    "react-test-renderer": "~16.8.6",
    "rimraf": "~2.6.3"
  },
  "peerDependencies": {
    "@babel/runtime": "^7.4.2",
    "react": "^16.8.0",
    "react-apollo": "^2.5.3",
    "react-dom": "^16.8.0",
    "react-router-dom": "~5.0.0"
  },
  "engines": {
    "node": "8.10.0 || >=10.7.0",
    "yarn": ">=1.12.0"
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
