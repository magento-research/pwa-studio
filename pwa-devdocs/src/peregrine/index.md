---
title: Peregrine
---

The Peregrine project is a collection of UI Components for Magento PWA projects.
Use, extend, or remix these components to create a unique Magento PWA storefront.

Visit the [`Peregrine`] repository to contribute to this project.

[`Peregrine`]: https://github.com/magento-research/pwa-studio/tree/master/packages/peregrine
