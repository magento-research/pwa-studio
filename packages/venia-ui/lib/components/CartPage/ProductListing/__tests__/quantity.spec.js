import React from 'react';
import { act } from 'react-test-renderer';
import { createTestInstance } from '@magento/peregrine';

import Quantity from '../quantity';

jest.mock('../../../../classify');
jest.mock('@apollo/react-hooks', () => {
    const executeMutation = jest.fn(() => ({ error: null }));
    const useMutation = jest.fn(() => [executeMutation]);

    return { useMutation };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

const mockOnChange = jest.fn();
const defaultProps = {
    itemId: 123,
    initialValue: 1,
    onChange: mockOnChange
};

test('renders quantity correctly', () => {
    const tree = createTestInstance(<Quantity {...defaultProps} />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('disables inc/dec buttons if no value', () => {
    const tree = createTestInstance(<Quantity {...defaultProps} />);

    tree.root.findByProps({ className: 'input' }).props.onChange({
        target: { value: undefined }
    });

    expect(
        tree.root.findByProps({ className: 'button_decrement' }).props.disabled
    ).toBe(true);
    expect(
        tree.root.findByProps({ className: 'button_increment' }).props.disabled
    ).toBe(true);
});

test('increments value on increment button click', () => {
    const tree = createTestInstance(<Quantity {...defaultProps} />);

    const oldValue = tree.root.findByProps({ className: 'input' }).props.value;

    act(() => {
        tree.root
            .findByProps({ className: 'button_increment' })
            .props.onClick();
    });

    const newValue = tree.root.findByProps({ className: 'input' }).props.value;
    expect(newValue).toBe(oldValue + 1);
    expect(mockOnChange).toHaveBeenNthCalledWith(1, oldValue + 1);
});

test('decrements value on decrement button click', () => {
    const props = {
        ...defaultProps,
        initialValue: 2
    };
    const tree = createTestInstance(<Quantity {...props} />);

    const oldValue = tree.root.findByProps({ className: 'input' }).props.value;

    act(() => {
        tree.root
            .findByProps({ className: 'button_decrement' })
            .props.onClick();
    });

    const newValue = tree.root.findByProps({ className: 'input' }).props.value;
    expect(newValue).toBe(oldValue - 1);
    expect(mockOnChange).toHaveBeenNthCalledWith(1, oldValue - 1);
});

test('input change restrains value to a min of 0', () => {
    const tree = createTestInstance(<Quantity {...defaultProps} />);

    act(() => {
        tree.root.findByProps({ className: 'input' }).props.onChange({
            target: { value: -1 }
        });
    });

    act(() => {
        expect(tree.root.findByProps({ className: 'input' }).props.value).toBe(
            0
        );
    });
});
