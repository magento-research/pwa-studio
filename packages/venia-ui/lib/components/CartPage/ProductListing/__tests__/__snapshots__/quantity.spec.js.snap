// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders quantity correctly 1`] = `
<div
  className="root"
>
  <div
    className="wrap"
  >
    <label
      className="label"
      htmlFor={123}
    >
      Quantity
    </label>
    <button
      aria-label="Decrease Quantity"
      className="button_decrement"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={22}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={22}
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </svg>
      </span>
    </button>
    <input
      className="input"
      id={123}
      onChange={[Function]}
      pattern="[0-9]*"
      type="text"
      value={1}
    />
    <button
      aria-label="Increase Quantity"
      className="button_increment"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        className="root"
      >
        <svg
          fill="none"
          height={22}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={22}
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="5"
            y2="19"
          />
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </svg>
      </span>
    </button>
  </div>
  <button
    className="updateButton"
    disabled={true}
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;
